Add undefined specializations in backward

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #43743 Update aliasing in tensorexpr fuser
* #43742 use types in the IR instead of vmap
* #43636 Add inline call to profiling executor pipeline
* #43635 Use prim::TensorExprGroup interned symbol
* #43634 Update requires grad property
* #43633 Specialize optionals for grad_sum_to_size
* **#43632 Add undefined specializations in backward**
* #43631 refactor tests
* #43630 Refactor pass to class
* #43629 Add API for unexecuted op

Specialize the backward graph by guarding on the undefinedness of the input tensors. The graph will look like:
```
ty1, ty2, succesful_checks = prim::TypeCheck(...)
if (succesful_checks)
-> optimized graph
else:
-> fallback graph
```


Specializing on the undefinedness of tensors allows us to clean up the 
```
if any_defined(inputs):
 outputs = <original_computation>
else:
 outputs = autograd zero tensors
```
blocks that make up the backward graph, so that we can fuse the original_computation nodes together.

Differential Revision: [D23358808](https://our.internmc.facebook.com/intern/diff/D23358808)

Commit:
Add undefined specializations in backward

[ghstack-poisoned]

Commit:
Update on "Add undefined specializations in backward"




[ghstack-poisoned]

ZolotukhinM commented on torch/csrc/jit/passes/specialize_autogradzero.cpp:
```diff
@@ -10,15 +10,49 @@ namespace jit {
+      return;
+    }
+    if (getProfilingMode()) {
+      if (auto vif = guardSpecializations()) {
```
Does `vif` stand for a "versioning if"? A comment could be helpful here I think.

ZolotukhinM commented on torch/csrc/jit/passes/specialize_autogradzero.cpp:
```diff
@@ -42,6 +76,98 @@ struct AutogradZeroSpecializer {
+      vif->destroy();
+      // the checks we inserted will be cleaned up
+      // by any subsequent DCE pass
+      return nullptr;
```
Nit: might be helpful to have a `GRAPH_DEBUG` statement here.

ZolotukhinM commented on torch/csrc/jit/passes/specialize_autogradzero.cpp:
```diff
@@ -10,15 +10,49 @@ namespace jit {
+  void replaceBlockInputsWithGraph(Block* b) {
+    TORCH_INTERNAL_ASSERT(graph_->inputs().size() == b->inputs().size());
+    for (int i = graph_->inputs().size() - 1; i >= 0; i--) {
+      b->inputs()[i]->replaceAllUsesWith(graph_->inputs()[i]);
```
Nit: maybe be do it in two loops to make clang-tidy happy.

ZolotukhinM reviewed (approved):
Looks good!

Krovatkin commented on torch/csrc/jit/passes/specialize_autogradzero.cpp:
```diff
@@ -42,6 +76,98 @@ struct AutogradZeroSpecializer {
+    Value* bool_list =
+        graph_->insertNode(graph_->createList(BoolType::get(), checks))
+            ->output();
+    Value* conjunction = graph_->insert(aten::all, {bool_list});
```
this is so much  more beautiful then a bunch of `ands`!!!

Krovatkin reviewed (approved):
:shipit:

Commit:
Update on "Add undefined specializations in backward"


Specialize the backward graph by guarding on the undefinedness of the input tensors. The graph will look like:
```
ty1, ty2, succesful_checks = prim::TypeCheck(...)
if (succesful_checks)
-> optimized graph
else:
-> fallback graph
```


Specializing on the undefinedness of tensors allows us to clean up the 
```
if any_defined(inputs):
 outputs = <original_computation>
else:
 outputs = autograd zero tensors
```
blocks that make up the backward graph, so that we can fuse the original_computation nodes together.

Differential Revision: [D23358808](https://our.internmc.facebook.com/intern/diff/D23358808)

[ghstack-poisoned]

Commit:
Update on "Add undefined specializations in backward"


Specialize the backward graph by guarding on the undefinedness of the input tensors. The graph will look like:
```
ty1, ty2, succesful_checks = prim::TypeCheck(...)
if (succesful_checks)
-> optimized graph
else:
-> fallback graph
```


Specializing on the undefinedness of tensors allows us to clean up the 
```
if any_defined(inputs):
 outputs = <original_computation>
else:
 outputs = autograd zero tensors
```
blocks that make up the backward graph, so that we can fuse the original_computation nodes together.

Differential Revision: [D23358808](https://our.internmc.facebook.com/intern/diff/D23358808)

[ghstack-poisoned]
